File: Plugins/Orimath.Basics/ViewModels/ToolListViewModel.fs
namespace Orimath.Basics.ViewModels
open System
open Mvvm
open Orimath.Plugins

type ToolListViewModel(workspace: IWorkspace, dispatcher: IDispatcher) as this =
    inherit NotifyPropertyChanged()
    let mutable tools = None
    do workspace.CurrentToolChanged.Add(fun _ ->
        ignore (dispatcher.OnUIAsync(Action(this.OnCurrentToolChanged))))

    member this.OnCurrentToolChanged() = this.OnPropertyChanged("CurrentTool")

    member __.Tools =
        match tools with
        | Some(result) -> result
        | None ->
            let result =
                workspace.Tools
                |> Seq.map(ToolViewModel)
                |> Seq.toArray
            tools <- Some(result)
            result

    member this.CurrentTool
        with get() =
            this.Tools
            |> Array.tryFind(fun vm -> vm.Source = workspace.CurrentTool)
            |> Option.defaultValue Unchecked.defaultof<_>
        and set(v: ToolViewModel) = workspace.CurrentTool <- v.Source

File: Plugins/Orimath.Basics.View/ViewModels/ToolListViewModel.cs
using System.Linq;
using Orimath.Controls;
using Orimath.Plugins;
using ApplicativeProperty;

namespace Orimath.Basics.View.ViewModels
{
    public class ToolListViewModel : NotifyPropertyChanged
    {
        private readonly IWorkspace _workspace;
        private readonly IDispatcher _dispatcher;

        private ToolViewModel[]? _tools;
        public ToolViewModel[] Tools => _tools ??= 
            _workspace.Tools.Select(t => new ToolViewModel(t)).ToArray();

        private ToolViewModel? _currentTool;
        public ToolViewModel? CurrentTool
        {
            get => _currentTool ??= Tools.FirstOrDefault(t => t.Source == _workspace.CurrentTool.Value);
            set => _workspace.CurrentTool.OnNext(value?.Source);
        }

        public ToolListViewModel(IWorkspace workspace, IDispatcher dispatcher)
        {
            _workspace = workspace;
            _dispatcher = dispatcher;
            _workspace.CurrentTool.Subscribe(_ => _dispatcher.OnUIAsync(() =>
            {
                _currentTool = null;
                OnPropertyChanged(nameof(CurrentTool));
            }));
        }
    }
}
